applications/StructuralMechanicsApplication/tests/cpp_tests/bits/adjoint_test_model_part_factory.cpp

```diff
@@ -0,0 +1,187 @@
+void CopySolutionStepData(ModelPart* pDestinationModelPart,
+                          const PointerVectorSet<Node<3>, IndexedObject>& rNodes);
+} // namespace smatmpf
+} // namespace
+
+ModelPart& CreateStructuralMechanicsAdjointTestModelPart(ModelPart* pPrimalModelPart)
+{
+    const std::string name = pPrimalModelPart->Name() + "(Adjoint)";
+    if (pPrimalModelPart->GetModel().HasModelPart(name))
+    {
+        pPrimalModelPart->GetModel().DeleteModelPart(name);
+    }
+    ModelPart& adjoint_model_part = pPrimalModelPart->GetModel().CreateModelPart(name);
+    smatmpf::AddVariables(&adjoint_model_part,
+                          pPrimalModelPart->GetNodalSolutionStepVariablesList());
+    smatmpf::CopyNodes(&adjoint_model_part, pPrimalModelPart->Nodes());
+    ProcessInfo& r_process_info = adjoint_model_part.GetProcessInfo();
+    r_process_info = pPrimalModelPart->GetProcessInfo();
+    smatmpf::CopyProperties(&adjoint_model_part, pPrimalModelPart->rProperties());
+    smatmpf::CreateAdjointElements(&adjoint_model_part, pPrimalModelPart->Elements());
+    adjoint_model_part.SetBufferSize(pPrimalModelPart->GetBufferSize());
+    // initialize adjoint time
+    adjoint_model_part.CloneTimeStep(r_process_info[TIME] + r_process_info[DELTA_TIME]);
+    smatmpf::AddDofs(&adjoint_model_part);
+    smatmpf::AssignBCs(&adjoint_model_part, *pPrimalModelPart);
+    VariableUtils().SetNonHistoricalVariable(UPDATE_SENSITIVITIES, true,
+                                             adjoint_model_part.Nodes());
+    smatmpf::CopySolutionStepData(&adjoint_model_part, pPrimalModelPart->Nodes());
+    return adjoint_model_part;
+}
+
+namespace
+{ // cpp internals
+namespace smatmpf
+{ // cotire unity guard
+void AddVariables(ModelPart* pAdjointModelPart, const VariablesList& rCustomVariables)
+{
+    pAdjointModelPart->GetNodalSolutionStepVariablesList() = rCustomVariables;
+    pAdjointModelPart->AddNodalSolutionStepVariable(ADJOINT_DISPLACEMENT);
+    pAdjointModelPart->AddNodalSolutionStepVariable(ADJOINT_VECTOR_2);
+    pAdjointModelPart->AddNodalSolutionStepVariable(ADJOINT_VECTOR_3);
+    pAdjointModelPart->AddNodalSolutionStepVariable(AUX_ADJOINT_VECTOR_1);
+    pAdjointModelPart->AddNodalSolutionStepVariable(SHAPE_SENSITIVITY);
+}
+
+void CopyNodes(ModelPart* pModelPart, const PointerVectorSet<Node<3>, IndexedObject>& rNodes)
+{
+    for (const auto& r_node : rNodes)
+    {
+        auto p_node = pModelPart->CreateNewNode(r_node.Id(), r_node.X0(),
+                                                r_node.Y0(), r_node.Z0());
+        p_node->Coordinates() = r_node.Coordinates();
+    }
+}
+
+void CopyProperties(ModelPart* pModelPart,
+                    const PointerVectorSet<Properties, IndexedObject>& rProperties)
+{
+    for (const auto& r_prop : rProperties)
+    {
+        pModelPart->GetProperties(r_prop.Id()) = r_prop;
+    }
+}
+
+const std::string* AdjointName(const std::string& rPrimalName)
```

Comment:
It avoids an allocation (not the most important here) since the length of these strings will most likely trigger one. I don't return a reference since there may be nothing to return (i.e., the string doesn't match). The pointer puts a trace into the function prototype to tell the developer this is optional.